mycpp/mylib.cc

```diff
@@ -208,6 +208,45 @@ Str* Str::join(List<Str*>* items) {
   return new Str(result, len);
 }
 
+
+  Str* Str::upper() {
+    Str* self = this;
+    Str* result = 0;
+
+    /* StackRoots stroots = {&result, &self}; */
+
+    result = mylib::BlankStr(self->len_);
+
+    // TODO(Jesse): Have to cast off const-ness .. this is intolerable in the
+    // long-run.  How should we deal?
+    char *buffer = (char*)result->data_;
+
+    for (int char_index = 0; char_index <= self->len_; ++char_index)
+    {
+      buffer[char_index] = toupper(self->data_[char_index]);
+    }
+    return result;
+  }
+
+  Str* Str::lower() {
+    Str* self = this;
+    Str* result = 0;
+
+    /* StackRoots stroots = {&result, &self}; */
+
+    result = mylib::BlankStr(self->len_);
+
+    // TODO(Jesse): Have to cast off const-ness .. this is intolerable in the
+    // long-run.  How should we deal?
+    char *buffer = (char*)result->data_;
+
+    for (int char_index = 0; char_index <= self->len_; ++char_index)
```

Comment:
this should be `< self->len_`, ditto above

Reply:
Ahh, right.  I was thinking I was being clever by copying nulls over (since the buffers are defined to be `len_+1`) but we don't actually have to do that. BlankStr() allocates an extra byte for us, which is 0'd out, correct?

I guess doing that extra iteration wouldn't hurt anything, but it's kinda sketchy if we ever change the defined size of the buffers, right?

I'm mostly just talking out loud here to confirm/deny my understanding of the memory layout.